sklearn/utils/_readonly_array_wrapper.pyx

```diff
@@ -0,0 +1,64 @@
+# cython: language_level=3
+
+# -------------------------------------
+# Readonly array wrapper
+# -------------------------------------
+# TODO: Remove with Cython >= 3.0 which supports const memoryviews for fused types.
+#
+# This class supports the buffer protocol, thus can wrap arrays and memoryviews.
+# All it does is LIE about the readonly attribute: tell it's false!
+# This way, we can use it on arrays that we don't touch.
+# !!! USE CAREFULLY !!!
+
+
+from cpython cimport Py_buffer
+from cpython.buffer cimport PyObject_GetBuffer, PyBuffer_Release, PyBUF_WRITABLE
+
+import numpy as np
+cimport numpy as np
+
+
+np.import_array()
+
+
+ctypedef fused NUM_TYPES:
+    np.npy_float64
+    np.npy_float32
+    np.npy_int64
+    np.npy_int32
```

Comment:
float should be identical to npy_float32 and double to npy_float64, see https://github.com/cython/cython/blob/bad93a3ab3b698ec4106dbf03e1a103e2582d003/Cython/Includes/numpy/__init__.pxd#L311-L312. I'll add the integers, though.

Reply:
I removed `int` and `long` because they gave a DeprecationWarning according to https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations. `np.npy_int64` and `np.npy_int32` should cover those anyway.